File: content/en/04-llms/llms/_exercises/15_evals/15_evals.R
library(ellmer)
library(vitals)

vitals::vitals_log_dir_set("./logs")

# Each case plants a subtle data-quality artifact in the prompt, and the
# eval grades whether each model flags the specific artifact unprompted.
source(here::here("_exercises/15_evals/_cases.R"))
cases <- bluff_mini_cases()

# The eval grades responses from two models against the same criteria.
models <- c(
  gemma = "google/gemma-4-26B-A4B-it",
  haiku = "claude-haiku-4-5"
)

# Run the eval once per model and collect pass rates by case
tasks <- purrr::imap(models, \(model, model_name) {
  tsk <- Task$new(
    dataset = cases,
    solver = generate(),
    scorer = model_graded_qa(
      scorer_chat = chat_posit(model = "claude-sonnet-5")
    ),
    name = model_name
  )

  tsk$eval(
    solver_chat = chat_posit(model = model),
    epochs = 2,
    view = FALSE
  )

  tsk
})

results <- purrr::map(tasks, \(tsk) {
  tsk$get_samples() |>
    dplyr::summarise(
      flagged = sum(score == "C"),
      n = dplyr::n(),
      pass_rate = mean(score == "C"),
      .by = id
    )
}) |>
  purrr::list_rbind(names_to = "model")

results

# Explore the transcripts in the log viewer
vitals::vitals_view()

File: content/en/04-llms/llms/_exercises/15_evals/15_evals.py
# %%
# Each case plants a subtle data-quality artifact in the prompt, and the
# eval grades whether each model flags the specific artifact unprompted.
import chatlas
import polars as pl
from _cases import bluff_mini_cases
from _setup import model_graded_qa_posit
from inspect_ai import Task, eval, view

cases = bluff_mini_cases()

# The eval grades responses from two models against the same criteria.
models = {
    "gemma": "google/gemma-4-26B-A4B-it",
    "haiku": "claude-haiku-4-5",
}

# %%
# Run the eval once per model
logs = []

for model_name, model in models.items():
    chat = chatlas.ChatPosit(model=model)
    tsk = Task(
        dataset=cases,
        solver=chat.to_solver(),
        scorer=model_graded_qa_posit(),
        name=f"bluff-mini-{model_name}",
        display_name=f"Bluff mini: {model_name}",
    )
    model_logs = eval(tsk, model=None, epochs=2, display="plain")
    logs.extend((model_name, log) for log in model_logs)

# %%
# Collect pass rates by case
rows = [
    {
        "model": model_name,
        "id": sample.id,
        "passed": sample.scores["model_graded_qa_posit"].value == "C",
    }
    for model_name, log in logs
    for sample in log.samples
]

results = (
    pl.DataFrame(rows)
    .group_by("model", "id", maintain_order=True)
    .agg(
        pl.col("passed").sum().alias("flagged"),
        pl.len().alias("n"),
        pl.col("passed").mean().alias("pass_rate"),
    )
)

results

# %%
# Explore the transcripts in the log viewer
view()
